File: benchmarks/nodejs/schema.js
const kuzu = require("kuzu");
const path = require("path");

/**
 * Create the authorization schema with nodes and relationships.
 * Implements a Zanzibar-inspired authorization model.
 */
async function createSchema(dbPath) {
  console.log("Creating KuzuDB schema for authorization...");

  const db = new kuzu.Database(dbPath);
  const conn = new kuzu.Connection(db);

  // Create User node table
  await conn.query(`
        CREATE NODE TABLE User(
            id STRING,
            name STRING,
            email STRING,
            created_at TIMESTAMP,
            metadata STRING,
            PRIMARY KEY (id)
        )
    `);
  console.log("✓ Created User node table");

  // Create Resource node table
  await conn.query(`
        CREATE NODE TABLE Resource(
            id STRING,
            type STRING,
            name STRING,
            owner_id STRING,
            created_at TIMESTAMP,
            metadata STRING,
            PRIMARY KEY (id)
        )
    `);
  console.log("✓ Created Resource node table");

  // Create Group node table
  await conn.query(`
        CREATE NODE TABLE UserGroup(
            id STRING,
            name STRING,
            description STRING,
            created_at TIMESTAMP,
            metadata STRING,
            PRIMARY KEY (id)
        )
    `);
  console.log("✓ Created Group node table");

  // Create MEMBER_OF relationship (User -> Group)
  await conn.query(`
        CREATE REL TABLE MEMBER_OF(
            FROM User TO UserGroup,
            joined_at TIMESTAMP,
            role STRING
        )
    `);
  console.log("✓ Created MEMBER_OF relationship table");

  // Create HAS_PERMISSION_USER relationship (User -> Resource)
  await conn.query(`
        CREATE REL TABLE HAS_PERMISSION_USER(
            FROM User TO Resource,
            can_create BOOLEAN,
            can_read BOOLEAN,
            can_update BOOLEAN,
            can_delete BOOLEAN,
            granted_at TIMESTAMP,
            granted_by STRING
        )
    `);
  console.log("✓ Created HAS_PERMISSION_USER relationship table");

  // Create HAS_PERMISSION_GROUP relationship (Group -> Resource)
  await conn.query(`
        CREATE REL TABLE HAS_PERMISSION_GROUP(
            FROM UserGroup TO Resource,
            can_create BOOLEAN,
            can_read BOOLEAN,
            can_update BOOLEAN,
            can_delete BOOLEAN,
            granted_at TIMESTAMP,
            granted_by STRING
        )
    `);
  console.log("✓ Created HAS_PERMISSION_GROUP relationship table");

  // Create INHERITS_FROM relationship (Group -> Group)
  await conn.query(`
        CREATE REL TABLE INHERITS_FROM(
            FROM UserGroup TO UserGroup,
            created_at TIMESTAMP
        )
    `);
  console.log("✓ Created INHERITS_FROM relationship table");

  console.log("\n✅ Schema created successfully!");

  return { db, conn };
}

async function printSchemaInfo(conn) {
  console.log("\n" + "=".repeat(60));
  console.log("Schema Information");
  console.log("=".repeat(60));

  const result = await conn.query("CALL SHOW_TABLES() RETURN *;");
  const tables = await result.getAll();

  console.log(`\nTotal tables: ${tables.length}`);
  tables.forEach((table) => {
    console.log(`  - ${table}`);
  });

  console.log("\nNode Tables:");
  console.log("  • User - system users");
  console.log("  • Resource - protected resources (documents, folders, etc.)");
  console.log("  • Group - user groups");

  console.log("\nRelationship Tables:");
  console.log("  • MEMBER_OF: User → Group");
  console.log("  • HAS_PERMISSION_USER: User → Resource");
  console.log("  • HAS_PERMISSION_GROUP: Group → Resource");
  console.log("  • INHERITS_FROM: Group → Group");

  console.log("\n" + "=".repeat(60));
}

async function main() {
  const fs = require("fs");

  // Database path
  const dbPath = path.join(__dirname, "..", "..", "db", "auth_test_nodejs");

  // Remove existing database if it exists
  if (fs.existsSync(dbPath)) {
    console.log(`⚠️  Removing existing database at ${dbPath}`);
    fs.rmSync(dbPath, { recursive: true, force: true });
  }

  // Create schema
  const { db, conn } = await createSchema(dbPath);

  // Print schema information
  await printSchemaInfo(conn);

  console.log(`\n✅ Database ready at: ${dbPath}`);

  // Close connection
  conn.close();
  db.close();
}

if (require.main === module) {
  main().catch((err) => {
    console.error("Error:", err);
    process.exit(1);
  });
}

module.exports = { createSchema, printSchemaInfo };

File: benchmarks/python/schema.py
"""
Schema definitions for KuzuDB authorization testing.
Implements a Zanzibar-inspired authorization model.
"""

import kuzu


def create_schema(db_path: str) -> kuzu.Database:
    """
    Create the authorization schema with nodes and relationships.

    Schema:
    - User nodes: represent users in the system
    - Resource nodes: represent resources (documents, folders, projects)
    - Group nodes: represent groups of users

    Relationships:
    - MEMBER_OF: User -> Group (user membership)
    - HAS_PERMISSION: User/Group -> Resource (direct permissions with CRUD flags)
    - INHERITS_FROM: Group -> Group (nested group hierarchies)

    Args:
        db_path: Path to the KuzuDB database directory

    Returns:
        Database instance with schema created
    """
    db = kuzu.Database(db_path)
    conn = kuzu.Connection(db)

    print("Creating KuzuDB schema for authorization...")

    # Create User node table
    conn.execute(
        """
        CREATE NODE TABLE User(
            id STRING,
            name STRING,
            email STRING,
            created_at TIMESTAMP,
            metadata STRING,
            PRIMARY KEY (id)
        )
    """
    )
    print("✓ Created User node table")

    # Create Resource node table
    conn.execute(
        """
        CREATE NODE TABLE Resource(
            id STRING,
            type STRING,
            name STRING,
            owner_id STRING,
            created_at TIMESTAMP,
            metadata STRING,
            PRIMARY KEY (id)
        )
    """
    )
    print("✓ Created Resource node table")

    # Create UserGroup node table
    conn.execute(
        """
        CREATE NODE TABLE UserGroup(
            id STRING,
            name STRING,
            description STRING,
            created_at TIMESTAMP,
            metadata STRING,
            PRIMARY KEY (id)
        )
    """
    )
    print("✓ Created UserGroup node table")

    # Create MEMBER_OF relationship (User -> UserGroup)
    conn.execute(
        """
        CREATE REL TABLE MEMBER_OF(
            FROM User TO UserGroup,
            joined_at TIMESTAMP,
            role STRING
        )
    """
    )
    print("✓ Created MEMBER_OF relationship table")

    # Create HAS_PERMISSION relationship (User -> Resource, Group -> Resource)
    # Permissions: create, read, update, delete (CRUD)
    conn.execute(
        """
        CREATE REL TABLE HAS_PERMISSION_USER(
            FROM User TO Resource,
            can_create BOOLEAN,
            can_read BOOLEAN,
            can_update BOOLEAN,
            can_delete BOOLEAN,
            granted_at TIMESTAMP,
            granted_by STRING
        )
    """
    )
    print("✓ Created HAS_PERMISSION_USER relationship table")

    conn.execute(
        """
        CREATE REL TABLE HAS_PERMISSION_GROUP(
            FROM UserGroup TO Resource,
            can_create BOOLEAN,
            can_read BOOLEAN,
            can_update BOOLEAN,
            can_delete BOOLEAN,
            granted_at TIMESTAMP,
            granted_by STRING
        )
    """
    )
    print("✓ Created HAS_PERMISSION_GROUP relationship table")

    # Create INHERITS_FROM relationship (UserGroup -> UserGroup)
    # For nested group hierarchies
    conn.execute(
        """
        CREATE REL TABLE INHERITS_FROM(
            FROM UserGroup TO UserGroup,
            created_at TIMESTAMP
        )
    """
    )
    print("✓ Created INHERITS_FROM relationship table")

    print("\n✅ Schema created successfully!")

    return db, conn


def print_schema_info(conn: kuzu.Connection):
    """Print information about the created schema."""
    print("\n" + "=" * 60)
    print("Schema Information")
    print("=" * 60)

    # Get node tables
    result = conn.execute("CALL SHOW_TABLES() RETURN *;")
    tables = []
    while result.has_next():
        tables.append(result.get_next())

    print(f"\nTotal tables: {len(tables)}")
    for table in tables:
        print(f"  - {table}")

    print("\nNode Tables:")
    print("  • User - system users")
    print("  • Resource - protected resources (documents, folders, etc.)")
    print("  • UserGroup - user groups")

    print("\nRelationship Tables:")
    print("  • MEMBER_OF: User → UserGroup")
    print("  • HAS_PERMISSION_USER: User → Resource")
    print("  • HAS_PERMISSION_GROUP: UserGroup → Resource")
    print("  • INHERITS_FROM: UserGroup → UserGroup")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    import sys
    import shutil
    from pathlib import Path

    # Database path
    db_path = Path(__file__).parent.parent / "db" / "auth_test"

    # Remove existing database if it exists
    if db_path.exists():
        print(f"⚠️  Removing existing database at {db_path}")
        shutil.rmtree(db_path)

    # Create schema
    db, conn = create_schema(str(db_path))

    # Print schema information
    print_schema_info(conn)

    print(f"\n✅ Database ready at: {db_path}")
